templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py

```diff
@@ -0,0 +1,65 @@
+"""
+Function call json extraction utils.
+Reference: https://gist.github.com/kouroshHakha/6cfbe2bf4aaafc5db733d408044f9902#file-create_test_dataset-py
+"""
+
+import json
+import re
+from fc_utils.preprocessing import TOOL_CALL_TAGS, TOOL_RESULT_TAGS
+
+class FunctionCallNotFoundError(Exception):
+    pass
+
+class FunctionResponseNotFoundError(Exception):
+    pass
+
+def extract_jsons(jsonl_string):
+    """
+    Extracts JSON objects from a string containing one or more JSONs.
+    Example: [{"name": "weather", "arguments": {\"location\": \"New York\"}}]
+    """
+    json_strs = []
+    brace_count = 0
+    start = None
+
+    for i, char in enumerate(jsonl_string):
+        if char == '{':
+            if brace_count == 0:
+                start = i
+            brace_count += 1
+        elif char == '}':
+            brace_count -= 1
+            if brace_count == 0:
+                end = i + 1
+                json_strs.append(jsonl_string[start:end])
+    jsons = [json.loads(json_str) for json_str in json_strs]
+    return jsons
+
+def get_tool_calls_from_response(assistant_content):
+    assistant_content = assistant_content.strip()  # remove trailing whitespaces
+    escaped_tool_call_tags = [re.escape(tag) for tag in TOOL_CALL_TAGS]
+    if assistant_content.startswith(TOOL_CALL_TAGS[0]):
```

Comment:
what is TOOL_CALL_TAGS[0]??

Reply:
Never hardcode things like this.

Reply:
Let's do some excercise (it will be a good one)

For each function you have defined, make high level docsring with all types hints and everything. Then feed the docstring and signature into AI and see how well it implemented it (catch its mistakes ofc, but I generally found its style much better than mine it will also give you better ideas on how to make code cleaner and easier to follow, because it was trained to do that). Overtime you'll pick up the right style.

Reply:
> Never hardcode things like this. 

"TOOL_CALL_TAGS" are the special indicators used during preprocessing ("[TOOL_CALLS]" and "[/TOOL_CALLS]") . Our post-processing code for the responses has to be pre-processing specific! There's no way to extract assistant response jsons if I don't know what tool call tags I used (assistant response can sometimes have valid json, but not a tool call).  

This is not buried deep within the code btw. It's at the top of [preprocessing.py](https://github.com/anyscale/templates/blob/b84ccda3ba3d22faa8a478c75d40526625ed3158/templates/fine-tune-function-calling/fc_utils/preprocessing.py#L11). This is one code file that has to have a dataset specific implementation. 

WDYT?

Reply:
> For each function you have defined, make high level docsring with all types hints and everything. Then feed the docstring and signature into AI and see how well it implemented it (catch its mistakes ofc, but I generally found its style much better than mine it will also give you better ideas on how to make code cleaner and easier to follow, because it was trained to do that). Overtime you'll pick up the right style.

Interesting. Let's see :)

Reply:
I see let's do something like this:

```
class ToolCallsTag(Enum):
       start = "[TOOL_CALLS]"
       end = "[/TOOL_CALLS]"
```

Reply:
I've put a higher-level NamedTuple called IndicatorTags and used that to instantiate our specific tool list, tool call and tool result tags.